### PR TITLE
Sematically ordered typesafe PackageVersion

### PR DESCRIPTION
**Pull request by tudor-da**

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala [hunk 1]
```diff
@@ -364,4 +365,40 @@ object Ref {
...
+    override def compare(that: PackageVersion): Int = {
+      import scala.math.Ordering.Implicits._
+
+      implicitly[Ordering[Seq[Int]]].compare(segments.toSeq, that.segments.toSeq)
+    }
+  }
+
+  object PackageVersion {
+    private val MaxPackageVersionLength = 1024
+    final def fromString(s: String): Either[String, PackageVersion] = for {
```
Comment (simonmaxen-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:384]: If we only use this for 1.16 packages that have this enforced by the compiler that is fine. If not then we need to think about a fallback. Maybe `ComparablePackageVersion`/`NonComparablePackageVersion` where the latter just stores a string (and gets rejected if used in a upgrading context).

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala [hunk 2]
```diff
@@ -364,4 +365,40 @@ object Ref {
...
+      _ <- Either.cond(
+        s.length <= MaxPackageVersionLength,
```
Comment (tudor-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:386]: We didn't have the size limit until now. Do you see a problem with introducing it now @remyhaemmerle-da ?

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala (hunk 1 above)
Comment (tudor-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:384]: Why would we differentiate? The only difference to the status quo is the size limit enforcement, which we can drop anyway.

File: daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala [hunk 3]
```diff
@@ -183,6 +183,55 @@ class RefTest extends AnyFreeSpec with Matchers with TableDrivenPropertyChecks w
...
+        "(Invalid) Raw version string",
+        "",
+        ".",
+        "1.",
+        ".0",
+        "00",
+        "1.002",
+        "0.-1",
+        "+1.0",
+        "beef.0",
```
Comment (carlpulley-da) [inline daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala:214]: Nitpick: maybe also test for something like "1.beef"?

Comment (carlpulley-da) [reply daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala:214]: As we also have a max limit, is there any value in testing that larger ints than this get rejected?

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM. :bow:

File: daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala:214]: Please add a test for the length checks. at least It should check 
- maxLentgth - 1, 
- maxLentgh,
-  maxLength + 1. 
- Very big string does not overflow the converter. (In particular length check should be done first).

Comment (remyhaemmerle-da) [reply daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala:214]: Can be done another PR.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:386]: I think limit  to 255 is fine.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:384]: I think there is an implicit limit already in the compiler. (packageName + packageVersion should be small than something). I have to double check with compiler guys.

File: daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala [hunk 4]
```diff
@@ -399,20 +474,20 @@ class RefTest extends AnyFreeSpec with Matchers with TableDrivenPropertyChecks w
         } ((x compare y) == 0) shouldBe (x == y)
       }
 
-      "is reflexive" - {
+      "is reflexive" in {
```
Comment (tudor-da) [inline daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala:477]: @remyhaemmerle-da These tests were not running. It's a pity that the `AnyFreeSpec` allows this. Enabled now

Comment (remyhaemmerle-da) [reply daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/RefTest.scala:478]: :bow:

**Review by remyhaemmerle-da — APPROVED** [review]

**Pull request merged**